File: server/src/main/java/com/classroom/server/service/MessageService.java
package com.classroom.server.service;

import com.classroom.server.dto.messages.SendMessageRequest;
import com.classroom.server.entity.Message;
import com.classroom.server.entity.User;

import java.util.List;

public interface MessageService {

    /**
     * Send new message or reply
     * - threadId == null → new thread
     * - threadId != null → reply
     */
    void send(Long courseId, User sender, SendMessageRequest request);

    /**
     * Inbox = one entry per thread (latest message of each thread)
     */
    List<Message> getThreadInbox(Long courseId, User user);

    /**
     * All messages (main + replies) of a thread
     */
    List<Message> getThreadMessages(Long courseId, Long threadId, User user);

    /**
     * Mark a single message as read (per recipient)
     */
    void markAsRead(Long messageId, User user);
}

File: client/src/services/messageService.js
import apiFetch from "./api";

export function sendMessage(courseId, userId, data) {
  return apiFetch(
    `/courses/${courseId}/messages?userId=${userId}`,
    {
      method: "POST",
      body: data,
    }
  );
}

export function getInbox(courseId, userId) {
  return apiFetch(
    `/courses/${courseId}/messages/inbox?userId=${userId}`
  );
}

export function getThreadMessages(courseId, threadId, userId) {
  return apiFetch(
    `/courses/${courseId}/messages/thread/${threadId}?userId=${userId}`
  );
}

export function markMessageAsRead(courseId, messageId, userId) {
  return apiFetch(
    `/courses/${courseId}/messages/${messageId}/read?userId=${userId}`,
    {
      method: "POST",
    }
  );
}
